packages/beacon-node/src/chain/regen/regen.ts

```diff
@@ -368,7 +385,31 @@ async function processSlotsToNearestCheckpoint(
     const cp = getCheckpointFromState(checkpointState);
     checkpointStateCache.add(cp, checkpointState);
     // consumers should not mutate or get the transfered cache
-    emitter.emit(ChainEvent.checkpoint, cp, checkpointState.clone(true));
+    emitter?.emit(ChainEvent.checkpoint, cp, checkpointState.clone(true));
+
+    if (count >= 1) {
+      // in normal condition, we only process 1 epoch so never reach this
+      // in that case, we want to prune state at the last 1/3 slot of slot 0 of the next epoch after importing the 1st block of epoch
+      // in non-finality time, we may process a lot of epochs so need to prune the cache to keep the node healthy
+      // this happened to holesky on Feb 2025, see https://github.com/ChainSafe/lodestar/issues/7495#issuecomment-2680800898
+      // cannot use getBlockRootAtSlot() because nextEpochSlot = postState
+      const latestBlockHex = toRootHex(cp.root);
+      try {
+        const persistCount = await checkpointStateCache.processState(latestBlockHex, checkpointState);
```

Comment:
Can we not split the logic to pruning cache and processing epoch? Mixing these both here is a bit confusing.

Reply:
One alternative would be to rely on the `ChainEvent.checkpoint` event, modifying it to take optionally pass in the checkpoint state.

And then somewhere 
```ts
onCheckpoint(cp: Checkpoint, checkpointState?: BeaconState) {
  if (!checkpointState) return;
  const latestBlockHex = toRootHex(cp.root);
  try {
    const persistCount = await checkpointStateCache.processState(latestBlockHex, checkpointState);
    logger?.verbose("pruning checkpointStateCache during processSlotsToNearestCheckpoint", {
        root: latestBlockHex,
        epoch: cp.epoch,
        persistCount,
      });
  } catch (e) {
    ...
  }
}
```

Although the difference is that in this PR, we await the pruning, and ^ we don't await.